Mathlib/RingTheory/MvPowerSeries/Evaluation.lean

```diff
@@ -0,0 +1,354 @@
+with `IsLinearTopology S S`, which means there is a basis of neighborhoods of 0
+consisting of ideals.
+
+Given `φ : R →+* S`, `a : σ → S`, and `f : MvPowerSeries σ R`,
+`MvPowerSeries.eval₂ f φ a` is the evaluation of the multivariate power series `f` at `a`.
+It `f` is (the coercion of) a polynomial, it coincides with the evaluation of that polynomial.
+Otherwise, it is defined by density from polynomials;
+its values are irrelevant unless `φ` is continuous and `a` satisfies two conditions
+bundled in `MvPowerSeries.HasEval a` :
+  - for all `s : σ`, `a s` is topologically nilpotent,
+    meaning that `(a s) ^ n` tends to 0 when `n` tends to infinity
+  - when `a s` tends to zero for the filter of cofinite subsets of `σ`.
+
+Under `Continuous φ` and `HasEval a`, the following lemmas furnish the properties of evaluation:
+
+* `MvPowerSeries.eval₂Hom`: the evaluation of multivariate power series, as a ring morphism,
+* `MvPowerSeries.aeval`: the evaluation map as an algebra morphism
+* `MvPowerSeries.uniformContinuous_eval₂`: uniform continuity of the evaluation
+* `MvPowerSeries.continuous_eval₂`: continuity of the evaluation
+* `MvPowerSeries.eval₂_eq_tsum`: the evaluation is given by the sum of its monomials, evaluated.
+
+-/
+
+namespace MvPowerSeries
+
+open Topology
+
+open Filter MvPolynomial RingHom Set TopologicalSpace UniformSpace
+
+/- ## Necessary conditions -/
+
+section
+
+variable {σ : Type*}
+variable {R : Type*} [CommRing R] [TopologicalSpace R]
+variable {S : Type*} [CommRing S] [TopologicalSpace S]
+variable {φ : R →+* S}
+
+-- We endow MvPowerSeries σ R with the Pi topology
+open WithPiTopology
+
+/-- Families at which power series can be consistently evaluated -/
+structure HasEval (a : σ → S) : Prop where
+  hpow : ∀ s, IsTopologicallyNilpotent (a s)
+  tendsto_zero : Tendsto a cofinite (𝓝 0)
+
+/-- The domain of evaluation of `MvPowerSeries`, as an ideal -/
+def hasEvalIdeal [IsTopologicalRing S] [IsLinearTopology S S] : Ideal (σ → S) where
+  carrier := {a | HasEval a}
+  add_mem' {a} {b} ha hb := {
+    hpow := fun s ↦ IsTopologicallyNilpotent.add (ha.hpow s) (hb.hpow s)
+    tendsto_zero := by
+      rw [← add_zero 0]
+      exact ha.tendsto_zero.add hb.tendsto_zero }
+  zero_mem' := {
+    hpow := fun s ↦ by
+      simp only [Pi.zero_apply]
+      apply tendsto_atTop_of_eventually_const (i₀ := 1)
+      intro i hi
+      rw [zero_pow (Nat.ne_zero_iff_zero_lt.mpr hi)]
+    tendsto_zero := tendsto_const_nhds }
+  smul_mem' c {x} hx := {
+    hpow := fun s ↦ by
+      simp only [IsTopologicallyNilpotent, Pi.smul_apply', smul_eq_mul, mul_pow]
+      exact IsLinearTopology.tendsto_mul_zero_of_right _ _ (hx.hpow s)
+    tendsto_zero := IsLinearTopology.tendsto_mul_zero_of_right _ _ hx.tendsto_zero }
+
+theorem HasEval.add [IsTopologicalRing S] [IsLinearTopology S S]
+    {a b : σ → S} (ha : HasEval a) (hb : HasEval b) : HasEval (a + b) :=
+  hasEvalIdeal.add_mem' ha hb
```

Comment:
nit: you should prove this and `zero` first, since it likely is more general than the ideal above

Reply:
@eric-wieser , I don't get your point: in what sense is this more general?

Reply:
For example, maybe this theorem holds in nonunital rings, where `Ideal` doesn't type-check. If you prove the theorem before doing the bundling, then it's easier for a future contributor to generalize it, or a future linter to auto detect this.

Reply:
OK, I did it.